### PR TITLE
Add `interpolate` to `resample_hours`

### LINKED ISSUE #2439: Interpolation in `resample_hours`
**Is your feature request related to a problem? Please describe.**
Our current [`resample_hours`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html#resample-hours) is only able to subset existing data. For example, it is able to create 3-hr data from 1-hr input data by only selecting every third time point.

It would be great to extend this with interpolation capabilities, so also time points that are not in the input data can be selected.

**Would you be able to help out?**
Yes!

### PR DESCRIPTION
**Pull request by schlunma**
## Description

Add interpolation capabilities to `resample_hours`.

Closes #2439

Link to documentation: https://esmvaltool--2440.org.readthedocs.build/projects/ESMValCore/en/2440/recipe/preprocessor.html#resample-hours

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Review by valeriupredoi — APPROVED** [review]
thanks Manu and Axel! Code review fine by me, but with a few very minor observations :beer:

File: doc/recipe/preprocessor.rst [hunk 1]
```diff
@@ -1570,29 +1570,43 @@ resample_hours:
...
-timesteps that belongs to the desired frequency. It is important to note that
-it is mainly mean to be used with instantaneous data
+Change the frequency of x-hourly data to y-hourly data by either eliminating
+extra time steps or interpolation.
+This is intended to be used with instantaneous data.
 
 Parameters:
-    * interval: New frequency of the data. Must be a divisor of 24
-    * offset: First desired hour. Default 0. Must be lower than the interval
+    * `interval`: New frequency of the data. Must be a divisor of 24.
```
[the remark this answers is not in the document]
Comment (schlunma) [reply doc/recipe/preprocessor.rst]: Done in https://github.com/ESMValGroup/ESMValCore/pull/2440/commits/81e7d74dbf830c0bdca9a8dc0a05721aa1282735

File: esmvalcore/preprocessor/_time.py [hunk 2]
```diff
@@ -1298,15 +1309,38 @@ def resample_hours(cube: Cube, interval: int, offset: int = 0) -> Cube:
...
+        else:
+            raise ValueError(
+                f"Expected `None`, 'nearest' or 'linear' for `interpolate`, "
+                f"got '{interpolate}'"
+            )
+        new_dates = sorted([
+            cf_datetime(y, m, d, h, calendar=time.units.calendar)
+            for h in range(0 + offset, 24, interval)
+            for (y, m, d) in {(d.year, d.month, d.day) for d in dates}
+        ])
```
[the remark this answers is not in the document]
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_time.py:1329]: thanks, Manu! I guess my PyOCD is not as bad as I thought :grin:

**Review by valeriupredoi — APPROVED** [review]
many thanks for the changes, Manu! This is RtM :beer:

**Review by alistairsellar — APPROVED** [review]

**Pull request merged**